[v2] Make the createApp API idiomatic for Express, fix service-worker loading.

olibrook opened the pull request:
# Description

@johnboxall Was looking at ways to support this pattern that we would have known from V1 (or any express app really):

```
app.get('/*', (req, res) => {
   res.set('cache-control', 'max-age=60')
   return render(req, res)
})
```

This is a reasonable request. I had accidentally buried the API that would make this possible behind the `addSSRRenderer` method. This PR makes the API explicit so that users like John can still set headers, or invoke rendering explicitly wherever they like.

I also found and fixed a critical issue that prevented service workers from loading properly (accidental re-name of a webpack compiler config, without checking for references).

## Testing

- Try out the snippet above and check that you can use it to set cache headers, in the way that John wanted.
- `curl localhost:3000/worker.js` should work on the retail template.
- `curl localhost:3000/worker.js.map` should work on the retail template.

Commit:
Make the createApp API explicit and fix service-worker loading bug.

bendvc commented on packages/pwa-kit-cli/src/ssr/server/build-dev-server.js:
```diff
@@ -173,6 +146,30 @@ export const DevServerMixin = {
+    render(req, res, next) {
+        const app = req.app
+        if (app.__webpackReady()) {
+            app.__hotServerMiddleware(req, res, next)
```
@vmarta This is the place that might conflict with your pr's change.

vmarta replied:
Thanks @bendvc for the heads up. My change is now simpler, so the merge conflict would be easier to resolve.

wjhsf commented on packages/pwa-kit-runtime/src/ssr/server/build-remote-server.js:
```diff
@@ -615,18 +618,94 @@ export const RemoteServerFactory = {
+        // Handled by the CDN on remote
+    },
+
+    addDevServerGc() {
```
```suggestion
    addDevServerGarbageCollection() {
```
`Gc` isn't very descriptive.

uip-robot-zz commented:
This issue has been linked to a new work item: [W-11050544](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE00000vk9lpYAA/view)

alexvuong commented on packages/pwa-kit-runtime/src/ssr/server/build-remote-server.js:
```diff
@@ -615,18 +618,94 @@ export const RemoteServerFactory = {
+     * a strong etag (to allow CDN-only revalidation), and to set maxage
+     * to 0 to prevent browser caching.
+     *
+     * See https://developers.google.com/web/updates/2018/06/fresher-sw
```
The site has been directed to https://developer.chrome.com/docs/android/trusted-web-activity/, which does not seem relevant to us

vmarta committed:
Merge branch 'v2' into improve-create-app-api

vcua-mobify commented on packages/pwa-kit-cli/src/ssr/server/build-dev-server.js:
```diff
@@ -173,6 +146,30 @@ export const DevServerMixin = {
+            const type = sourceMap ? '.js.map' : '.js'
+            const content = DevServerFactory._getWebpackAsset(req, 'client-optional', file)
+            if (content === null) {
+                res.sendStatus(404)
```
Iirc, waitUntilValid should pause the request until webpack is able to serve the service worker. Wouldn't returning a 404 undermine that?

olibrook replied:
No. If `content === null` then the user hasn't created a service worker. There is nothing sensible that you _could_ serve in its place, even if you wanted to.

Reminder that service workers are optional for MRT apps, but very important for the Retail React App.

Commit:
Update docstring on service-worker handling - Alex's feedback

Commit:
Rename addDevServerGc -> addDevServerGarbageCollection - Will's feedback

Commit:
Merge branch 'improve-create-app-api' of https://github.com/SalesforceCommerceCloud/pwa-kit into improve-create-app-api

vmarta committed:
Remove mentions of the old `addSsrRenderer`

vmarta committed:
Add comment to clarify

vmarta committed:
Use more-specific routes

vmarta commented on packages/pwa-kit-cli/src/ssr/server/build-dev-server.js:
```diff
@@ -100,6 +99,7 @@ export const DevServerMixin = {
                 console.log(chalk.cyan('First build complete'))
             }, 75)
         })
+        app.__hotServerMiddleware = webpackHotServerMiddleware(app.__compiler)
```
Moving this line from the old `addSSRRenderer` to here in `addSDKInternalHandlers` makes it possible to break the local dev server. Seeing this error when `npm start` from within the template-express-minimal:

```
/Users/vmarta/repos/pwa-kit-copy/packages/pwa-kit-cli/node_modules/webpack-hot-server-middleware/src/index.js:117
        throw new Error(`Expected a webpack compiler named 'server'`);
```

Why is the error happening? Because now webpack hot server middleware is attached for every template, while it used to be attached _only_ when `addSSRRenderer()` is called.

olibrook replied:
Incredibly well spotted – thank you so much!

vmarta committed:
Attach the hot server middleware if necessary

If a project does not require pwa-kit-react-sdk, then the webpack config won't include the 'server' config. Thus, the hot server middleware, which needs that server config, should not be used.

olibrook commented on packages/pwa-kit-cli/src/ssr/server/build-dev-server.js:
```diff
@@ -100,6 +99,9 @@ export const DevServerMixin = {
         })
+        if (config.some((cnf) => cnf.name === 'server')) {
+            app.__hotServerMiddleware = webpackHotServerMiddleware(app.__compiler)
+        }
```
We haven't communicated the dependencies on these compiler names very well. But, I'm not sure I can see a fix that is better than this right now.

It'd be great if we could remain open to swapping the HMR implementation out in future! [Discussion here](https://mobify.slack.com/archives/C0DKK1FJS/p1651076799364259).

vmarta commented:
@olibrook When I tried to set the cache-control header (just as what John had wanted to do), I saw that I needed to restart the dev server so that the changes to ssr.js can be seen. We didn't have to restart it in V1, but here in V2 (with the different dev server used) we need to restart the server to test the code changes in ssr.js.

How can we make it easier for developers to test their ssr.js changes? 🤔

vmarta committed:
Merge branch 'v2' into improve-create-app-api

vmarta committed:
Use more specific route

vmarta committed:
Add a new test for react rendering

vmarta committed:
Refactor the hardcoded webpack config names

vmarta commented on packages/template-retail-react-app/app/ssr.js:
```diff
@@ -63,7 +63,8 @@ const {handler} = runtime.createHandler(options, (app) => {
     app.get('/robots.txt', serveStaticFile('static/robots.txt'))
 
-    runtime.addSSRRenderer(app)
+    app.get('/worker.js(.map)?', runtime.serveServiceWorker)
+    app.get('*', runtime.render)
```
@kevinxh (cc @alexvuong ) Getting back to our conversation from yesterday: specifically, why is `serveServiceWorker` a method from the `runtime` object (while `serveStaticFile` on line 64 is not)? One reason it's because:

- `serveServiceWorker` has different implementation depending on whether the runtime is [local dev](https://github.com/SalesforceCommerceCloud/pwa-kit/blob/eb86d1970912a230696aa7e4a4d9ae57408aa809/packages/pwa-kit-cli/src/ssr/server/build-dev-server.js#L151-L166) or [remote](https://github.com/SalesforceCommerceCloud/pwa-kit/blob/eb86d1970912a230696aa7e4a4d9ae57408aa809/packages/pwa-kit-runtime/src/ssr/server/build-remote-server.js#L643-L667) server
- `serveStaticFile` is a function that has the _same_ implementation regardless of local or remote server.

kevinxh replied:
From an end user's perspective, it would make sense to me that all the http serving functionalities are from the same place, which is the runtime. I think we should design the API to be more user friendly as oppose to based on the actual implementation.

olibrook replied:
> serveStaticFile is a function that has the same implementation regardless of local or remote serve

There are other things that are the same on dev/prod, too, and we still include these in the runtime. The decision to include it or not isn't based on this logic – it's simply that I didn't have time/feedback to move it.

I am with @kevinxh on this, and I had the same thought myself as I looked at it this week.

kevinxh replied:
On a similar note, IMO utilities like `isRemote` should be a part of the runtime API, what's your thought on moving `isRemote` to the runtime? @olibrook @vmarta

kevinxh replied:
The thing that we are missing is a way to make `RemoteServerFactory` and `DevServerFactory` to inherit a base class, or an abstract factory.

olibrook replied:
Well, I would avoid opening a can of worms here. There are lots and lots of things that you could include. But it's a lil' late in the day. The `serveStaticFile` thing is dead obvious. I think `isRemote` is less obvious, because you want to use it in React code, too (and you typically don't access the runtime from React).

> The thing that we are missing is a way to make RemoteServerFactory and DevServerFactory to inherit a base class, or an abstract factory.

We're not missing this – what we have here is something I arrived at after having tried those already and then very carefully pulling apart the dependencies to make sure we don't accidentally ship devServer to production.

I would (1) move `serveStaticFile` because it's easy and has nice aesthetics (2) ship V2 and (3) think about some simple API changes that you could do for a V3 release – we should get less sensitive about major version bumps.

@kevinxh Your time would be better spent doing that hot-reloading work you were excited about, no?

bendvc replied:
Agreed. Great idea @kevinxh. As much as I hate saying it, its probably too late to be doing a refactor like that. We could however make a ticket/doc to track that so that idea isn't lost and aim for v3 on that one like Oli said. But moving `serveStaticFile` makes sense now.

kevinxh replied:
👍

vmarta replied:
All right, `serveStaticFile` has been moved.

vmarta committed:
Merge branch 'v2' into improve-create-app-api

vmarta committed:
Move serveStaticFile into the remote runtime

For a more consistent API.

vmarta committed:
Fix lint errors

vmarta committed:
Merge branch 'v2' into improve-create-app-api
